core/src/main/java/org/apache/iceberg/rest/RESTFileScanTaskParser.java

```diff
@@ -0,0 +1,109 @@
+  private static final String RESIDUAL = "residual-filter";
+
+  private RESTFileScanTaskParser() {}
+
+  public static void toJson(
+      FileScanTask fileScanTask, List<DeleteFile> deleteFiles, JsonGenerator generator)
+      throws IOException {
+    Preconditions.checkArgument(fileScanTask != null, "Invalid file scan task: null");
+    Preconditions.checkArgument(generator != null, "Invalid JSON generator: null");
+
+    generator.writeStartObject();
+    generator.writeFieldName(DATA_FILE);
+    RESTContentFileParser.toJson(fileScanTask.file(), generator);
+
+    // TODO revisit this logic
+    if (deleteFiles != null) {
+      generator.writeArrayFieldStart(DELETE_FILE_REFERENCES);
+      for (int delIndex = 0; delIndex < deleteFiles.size(); delIndex++) {
+        generator.writeNumber(delIndex);
+      }
+      generator.writeEndArray();
+    }
+    if (fileScanTask.residual() != null) {
+      generator.writeFieldName(RESIDUAL);
+      ExpressionParser.toJson(fileScanTask.residual(), generator);
+    }
+    generator.writeEndObject();
+  }
+
+  public static FileScanTask fromJson(JsonNode jsonNode, List<DeleteFile> allDeleteFiles) {
+    Preconditions.checkArgument(jsonNode != null, "Invalid JSON node for file scan task: null");
+    Preconditions.checkArgument(
+        jsonNode.isObject(), "Invalid JSON node for file scan task: non-object (%s)", jsonNode);
+
+    DataFile dataFile = (DataFile) RESTContentFileParser.fromJson(jsonNode.get(DATA_FILE));
+
+    DeleteFile[] matchedDeleteFiles = null;
+    List<Integer> deleteFileReferences = null;
+    if (jsonNode.has(DELETE_FILE_REFERENCES)) {
+      ImmutableList.Builder deleteFileReferencesBuilder = ImmutableList.builder();
+      JsonNode deletesArray = jsonNode.get(DELETE_FILE_REFERENCES);
+      for (JsonNode deleteRef : deletesArray) {
+        deleteFileReferencesBuilder.add(deleteRef);
+      }
+      deleteFileReferences = deleteFileReferencesBuilder.build();
+    }
+
+    if (deleteFileReferences != null) {
+      ImmutableList.Builder matchedDeleteFilesBuilder = ImmutableList.builder();
+      for (Integer deleteFileIdx : deleteFileReferences) {
+        matchedDeleteFilesBuilder.add(allDeleteFiles.get(deleteFileIdx));
+      }
+      matchedDeleteFiles = (DeleteFile[]) matchedDeleteFilesBuilder.build().stream().toArray();
+    }
+
+    // TODO revisit this in spec
+    Expression filter = Expressions.alwaysTrue();
+    if (jsonNode.has(RESIDUAL)) {
+      filter = ExpressionParser.fromJson(jsonNode.get(RESIDUAL));
+    }
+
+    ResidualEvaluator residualEvaluator = ResidualEvaluator.of(filter);
+
+    // TODO at the time of creation we dont have the schemaString and specString so can we avoid
+    // setting this
+    // will need to refresh before returning closed iterable of tasks, for now put place holder null
+    BaseFileScanTask baseFileScanTask =
+        new BaseFileScanTask(dataFile, matchedDeleteFiles, null, null, residualEvaluator);
```

Comment:
we can introduce a `UnboundFileScanTask` which extends BaseFileScanTask and overrides the spec/schema to throw Unsupported. That class could have a `bind(PartitionSpec spec, Schema schema)` API which returns a BaseFileScanTask with the updated spec/schema, and the data file partition data updated correctly.

Reply:
Will try this idea out thanks again @amogh-jahagirdar !

Reply:
I think we also need to bind the residual as well.

Reply:
[doubt] These fileScanTasks can be belong to diff snapshots and we would have to bind this this UnboundFileScanTask to get FileScanTask via first getting the schema of the snapshot this fileScan task belongs to then get PartitionSpec from binding that schema with the UnboundedPartitionSpec and then we get to it. Do we have a metadata for mapping FileScanTask to a snapshot id ? 
ref this interesting case : https://github.com/apache/iceberg/issues/11162